ws/api/endpoints.go

```diff
@@ -0,0 +1,186 @@
+			return connReq{}, errUnauthorizedAccess
+		}
+		authKey = authKeys[0]
+	}
+
+	chanID := bone.GetValue(r, "id")
+
+	req := connReq{
+		thingKey: authKey,
+		chanID:   chanID,
+	}
+
+	channelParts := channelPartRegExp.FindStringSubmatch(r.RequestURI)
+	if len(channelParts) < 2 {
+		logger.Warn("Empty channel id or malformed url")
+		return connReq{}, errors.ErrMalformedEntity
+	}
+
+	subtopic, err := parseSubTopic(channelParts[2])
+	if err != nil {
+		return connReq{}, err
+	}
+
+	req.subtopic = subtopic
+
+	return req, nil
+}
+
+func parseSubTopic(subtopic string) (string, error) {
+	if subtopic == "" {
+		return subtopic, nil
+	}
+
+	subtopic, err := url.QueryUnescape(subtopic)
+	if err != nil {
+		return "", errMalformedSubtopic
+	}
+
+	subtopic = strings.Replace(subtopic, "/", ".", -1)
+
+	elems := strings.Split(subtopic, ".")
+	filteredElems := []string{}
+	for _, elem := range elems {
+		if elem == "" {
+			continue
+		}
+
+		if len(elem) > 1 && (strings.Contains(elem, "*") || strings.Contains(elem, ">")) {
+			return "", errMalformedSubtopic
+		}
+
+		filteredElems = append(filteredElems, elem)
+	}
+
+	subtopic = strings.Join(filteredElems, ".")
+
+	return subtopic, nil
+}
+
+func listen(conn *websocket.Conn, msgs chan<- []byte) {
+	for {
+		// Listen for message from the client, and push them to the msgs channel
+		_, payload, err := conn.ReadMessage()
+
+		if websocket.IsUnexpectedCloseError(err) {
+			logger.Debug(fmt.Sprintf("Closing WS connection: %s", err.Error()))
+			close(msgs)
+			return
+		}
+
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to read message: %s", err.Error()))
+			close(msgs)
+			return
+		}
+
+		msgs <- payload
+	}
+}
+
+func process(svc ws.Service, req *connReq, msgs <-chan []byte) {
+	for msg := range msgs {
+		m := messaging.Message{
+			Channel:  req.chanID,
+			Subtopic: req.subtopic,
+			Protocol: "websocket",
+			Payload:  msg,
+			Created:  time.Now().UnixNano(),
+		}
+		svc.Publish(context.Background(), req.thingKey, m)
+	}
+	if err := svc.Unsubscribe(context.Background(), req.thingKey, req.chanID, req.subtopic); err != nil {
+		logger.Warn(fmt.Sprintf("Failed to subscribe to broker: %s", err.Error()))
+		req.conn.Close()
+	}
+}
+
+func encodeError(req *connReq, w http.ResponseWriter, err error) {
+	statusCode := http.StatusUnauthorized
+
+	switch err {
+	case ws.ErrEmptyID, ws.ErrEmptyTopic:
+		statusCode = http.StatusBadRequest
+	case errUnauthorizedAccess:
+		statusCode = http.StatusForbidden
+	case errMalformedSubtopic, errors.ErrMalformedEntity:
+		statusCode = http.StatusBadRequest
+	default:
+		switch e, ok := status.FromError(err); {
+		case ok:
+			switch e.Code() {
+			case codes.Unauthenticated:
+				statusCode = http.StatusUnauthorized
+			case codes.PermissionDenied:
+				statusCode = http.StatusForbidden
+			case codes.Internal:
+				statusCode = http.StatusInternalServerError
+			default:
+				statusCode = http.StatusInternalServerError
+			}
+		}
+	}
```

Comment:
These cases are not covered by tests. Since gRPC calls errors are wrapped by `errors.ErrAuthorization` on the adapter (service) level - it looks like gRPC errors will never happen anyways - so we don't need these cases. Please double-check.